### ISSUE TITLE
Crashes on hovering over anything in the source file

### ISSUE THREAD

**Issue by Lucretia** [AdaCore/ada_language_server#202]
Move the mouse over the source, in the output pane get:

```
[Error - 16:34:15] Request textDocument/documentSymbol failed.
  Message: CONSTRAINT_ERROR
[/home/laguest/.vscode-insiders/extensions/adacore.ada-0.0.1/linux/libgnatcoll.so.0]
0x7fa97bfd9986 gnatcoll__projects__info_set at ???
[/home/laguest/.vscode-insiders/extensions/adacore.ada-0.0.1/linux/ada_language_server]
Lsp.Ada_Contexts.Is_Part_Of_Project at lsp-ada_contexts.adb:241
Lsp.Ada_Handlers.Get_Best_Context_For_File at lsp-ada_handlers.adb:135
Lsp.Ada_Handlers.Get_Best_Context_For_Uri at lsp-ada_handlers.adb:157
Lsp.Ada_Handlers.Get_Document at lsp-ada_handlers.adb:170
Lsp.Ada_Handlers.On_Document_Symbols_Request at lsp-ada_handlers.adb:1297
Lsp.Servers.Handle_Request at lsp-servers-handle_request.adb:138
Lsp.Servers.Processing_Task_Type.Process_Message at lsp-servers.adb:841
Lsp.Servers.Processing_Task_TypeT at lsp-servers.adb:923
[/home/laguest/.vscode-insiders/extensions/adacore.ada-0.0.1/linux/libgnarl-2019.so]
0x7fa97a0043f4 system__tasking__stages__task_wrapper at ???
[/lib64/libpthread.so.0]
0x7fa9799883a5
[/lib64/libc.so.6]
0x7fa97988b89d
0xfffffffffffffffe

  Code: -32603 
[Error - 16:34:15] Request textDocument/hover failed.
  Message: CONSTRAINT_ERROR
[/home/laguest/.vscode-insiders/extensions/adacore.ada-0.0.1/linux/libgnatcoll.so.0]
0x7fa97bfd9986 gnatcoll__projects__info_set at ???
[/home/laguest/.vscode-insiders/extensions/adacore.ada-0.0.1/linux/ada_language_server]
Lsp.Ada_Contexts.Is_Part_Of_Project at lsp-ada_contexts.adb:241
Lsp.Ada_Handlers.Get_Best_Context_For_File at lsp-ada_handlers.adb:135
Lsp.Ada_Handlers.Get_Best_Context_For_Uri at lsp-ada_handlers.adb:157
Lsp.Ada_Handlers.On_Hover_Request at lsp-ada_handlers.adb:1025
Lsp.Servers.Handle_Request at lsp-servers-handle_request.adb:102
Lsp.Servers.Processing_Task_Type.Process_Message at lsp-servers.adb:841
Lsp.Servers.Processing_Task_TypeT at lsp-servers.adb:923
[/home/laguest/.vscode-insiders/extensions/adacore.ada-0.0.1/linux/libgnarl-2019.so]
0x7fa97a0043f4 system__tasking__stages__task_wrapper at ???
[/lib64/libpthread.so.0]
0x7fa9799883a5
[/lib64/libc.so.6]
0x7fa97988b89d
0xfffffffffffffffe

  Code: -32603 
```

**Comment by Lucretia** [comment]
Ok, installing latest version seems fine.

**Comment by Lucretia** [comment]
Just started vscode and it happened again, I've not changed anything with the plugin since yesterday.

**Comment by Lucretia** [comment]
I did that when Maxim asked me to, it seemed fine after I installed it, then when I started vscode again, it started crashing again.

**Comment by setton** [comment]
Do you have a project / scenario that I can try on my end?

If your last attempt was in September, I would strongly recommend upgrading to today's version - we've made a lot of progress.

**Comment by setton** [comment]
Hi @Lucretia - it looks like it works for me.

I have this simple settings file:
```
$ cat .vscode/settings.json 
{ "ada.projectFile": "./build/gnat/sdlada.gpr" }
```

And it looks to me that the contents of the hover is "as expected" - see this screenshot. Let me know if I've missed something!

<img width="1335" alt="Screenshot 2019-11-07 at 13 12 50" src="https://user-images.githubusercontent.com/96973/68391940-91a1a700-0160-11ea-82b6-ea7d04ee241e.png">

**Comment by setton** [comment]
(it works from sdl-video-surfaces.ads:59:67 as well)

**Comment by Lucretia** [comment]
Don't know what I did, but the problems window just came up with this:

```bash
[Error - 20:55:30] Request textDocument/documentSymbol failed.
  Message: SYSTEM.ASSERTIONS.ASSERT_FAILURE
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.1/linux/libgnat-2019.so]
0x7f51dcd25bc1
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.1/linux/ada_language_server]
0x6347d3
0x660b7a
0x674635
0x5d8684
0x5da901
0x5e9fbb
0x609c2e
0x609425
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.1/linux/libgnarl-2019.so]
0x7f51dd08c3f4
[/lib64/libpthread.so.0]
0x7f51dca082e5
[/lib64/libc.so.6]
0x7f51dc90df6d
0xfffffffffffffffe

  Code: -32603 
[Error - 20:55:31] Request textDocument/documentSymbol failed.
  Message: SYSTEM.ASSERTIONS.ASSERT_FAILURE
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.1/linux/libgnat-2019.so]
0x7f51dcd25bc1
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.1/linux/ada_language_server]
0x6347d3
0x660b7a
0x674635
0x5d8684
0x5da901
0x5e9fbb
0x609c2e
0x609425
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.1/linux/libgnarl-2019.so]
0x7f51dd08c3f4
[/lib64/libpthread.so.0]
0x7f51dca082e5
[/lib64/libc.so.6]
0x7f51dc90df6d
0xfffffffffffffffe

  Code: -32603 
[Error - 20:55:31] Request textDocument/documentSymbol failed.
  Message: SYSTEM.ASSERTIONS.ASSERT_FAILURE
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.1/linux/libgnat-2019.so]
0x7f51dcd25bc1
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.1/linux/ada_language_server]
0x6347d3
0x660b7a
0x674635
0x5d8684
0x5da901
0x5e9fbb
0x609c2e
0x609425
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.1/linux/libgnarl-2019.so]
0x7f51dd08c3f4
[/lib64/libpthread.so.0]
0x7f51dca082e5
[/lib64/libc.so.6]
0x7f51dc90df6d
0xfffffffffffffffe

  Code: -32603 
[Error - 20:55:32] Request textDocument/documentSymbol failed.
  Message: SYSTEM.ASSERTIONS.ASSERT_FAILURE
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.1/linux/libgnat-2019.so]
0x7f51dcd25bc1
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.1/linux/ada_language_server]
0x6347d3
0x660b7a
0x674635
0x5d8684
0x5da901
0x5e9fbb
0x609c2e
0x609425
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.1/linux/libgnarl-2019.so]
0x7f51dd08c3f4
[/lib64/libpthread.so.0]
0x7f51dca082e5
[/lib64/libc.so.6]
0x7f51dc90df6d
0xfffffffffffffffe

  Code: -32603 
```

**Comment by setton** [comment]
Hmm, looks like we need to fiddle with the switches so that we can symbolize these backtraces - or keep an executable with debug info on the side so we can do that post-mortem... @reznikmm what do you think?

**Comment by c-grant** [comment]
I get a similar message when hovering over a package of a user defined type

[c:\Users\<redacted>\.vscode\extensions\maximreznik.ada-21.0.4\win32\libgnat-2019\libgnat-2019.dll]
0x6a7f3386
[c:\Users\<redacted>\.vscode\extensions\maximreznik.ada-21.0.4\win32\libadalang\libadalang.dll]
0x20bd6d5
[c:\Users\<redacted>\.vscode\extensions\maximreznik.ada-21.0.4\win32\libadalang\libadalang.dll]
0x20cb473
[c:\Users\<redacted>\.vscode\extensions\maximreznik.ada-21.0.4\win32\ada_language_server.exe]
0x82356a
0x759417
0x75d70c
0x7675a8
0x791fd1
0x790f7b
[c:\Users\<redacted>\.vscode\extensions\maximreznik.ada-21.0.4\win32\libgnarl-2019\libgnarl-2019.dll]
0x712d43bc
[C:\WINDOWS\System32\KERNEL32.DLL]
0x7ff8e2d27972
[C:\WINDOWS\SYSTEM32\ntdll.dll]
0x7ff8e55aa26f

  Code: -32603

**Comment by reznikmm** [comment]
@Lucretia Could you try to replace ALS with debug version [from here](https://bintray.com/reznikmm/ada-language-server/download_file?file_path=linux-21.0.4-dbg.tar.gz)
to get symbolic traceback?

**Comment by reznikmm** [comment]
@c-grant Can you provide a reproducer? (like source of the package and positon of the cursor)

**Comment by Lucretia** [comment]
```
[Error - 17:15:36] Request textDocument/completion failed.
  Message: CONSTRAINT_ERROR
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.4/linux/libadalang.so]
Libadalang.Analysis.Kind at libadalang-analysis.adb:10690
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.4/linux/ada_language_server]
Lsp.Ada_Documents.Get_Completions_At at lsp-ada_documents.adb:495
Lsp.Ada_Handlers.On_Completion_Request at lsp-ada_handlers.adb:2313
Lsp.Error_Decorators.Completion_Request at lsp-error_decorators.adb:45
Lsp.Error_Decorators.On_Completion_Request at lsp-error_decorators.adb:117
Lsp.Servers.Handle_Request at lsp-servers-handle_request.adb:55
Lsp.Servers.Processing_Task_Type.Process_Message at lsp-servers.adb:1012
Lsp.Servers.Processing_Task_TypeT at lsp-servers.adb:1089
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.4/linux/libgnarl-2019.so]
0x7f6b2e3143f4 system__tasking__stages__task_wrapper at ???
[/lib64/libpthread.so.0]
0x7f6b2dc882e5
[/lib64/libc.so.6]
0x7f6b2db8dfed
0xfffffffffffffffe

  Code: -32603 
[Error - 17:15:36] Request textDocument/completion failed.
  Message: CONSTRAINT_ERROR
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.4/linux/libadalang.so]
Libadalang.Analysis.Kind at libadalang-analysis.adb:10690
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.4/linux/ada_language_server]
Lsp.Ada_Documents.Get_Completions_At at lsp-ada_documents.adb:495
Lsp.Ada_Handlers.On_Completion_Request at lsp-ada_handlers.adb:2313
Lsp.Error_Decorators.Completion_Request at lsp-error_decorators.adb:45
Lsp.Error_Decorators.On_Completion_Request at lsp-error_decorators.adb:117
Lsp.Servers.Handle_Request at lsp-servers-handle_request.adb:55
Lsp.Servers.Processing_Task_Type.Process_Message at lsp-servers.adb:1012
Lsp.Servers.Processing_Task_TypeT at lsp-servers.adb:1089
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.4/linux/libgnarl-2019.so]
0x7f6b2e3143f4 system__tasking__stages__task_wrapper at ???
[/lib64/libpthread.so.0]
0x7f6b2dc882e5
[/lib64/libc.so.6]
0x7f6b2db8dfed
0xfffffffffffffffe

  Code: -32603 
[Error - 17:15:39] Request textDocument/completion failed.
  Message: CONSTRAINT_ERROR
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.4/linux/libadalang.so]
Libadalang.Analysis.Kind at libadalang-analysis.adb:10690
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.4/linux/ada_language_server]
Lsp.Ada_Documents.Get_Completions_At at lsp-ada_documents.adb:495
Lsp.Ada_Handlers.On_Completion_Request at lsp-ada_handlers.adb:2313
Lsp.Error_Decorators.Completion_Request at lsp-error_decorators.adb:45
Lsp.Error_Decorators.On_Completion_Request at lsp-error_decorators.adb:117
Lsp.Servers.Handle_Request at lsp-servers-handle_request.adb:55
Lsp.Servers.Processing_Task_Type.Process_Message at lsp-servers.adb:1012
Lsp.Servers.Processing_Task_TypeT at lsp-servers.adb:1089
[/home/laguest/.vscode-insiders/extensions/maximreznik.ada-21.0.4/linux/libgnarl-2019.so]
0x7f6b2e3143f4 system__tasking__stages__task_wrapper at ???
[/lib64/libpthread.so.0]
0x7f6b2dc882e5
[/lib64/libc.so.6]
0x7f6b2db8dfed
0xfffffffffffffffe

  Code: -32603 
```

**Comment by c-grant** [comment]
Attached are a representative set of file that make it happen. 
[test.zip](https://github.com/AdaCore/ada_language_server/files/4161404/test.zip)

To make it happen hover over "Test" on line 9 of test2.ads

**Comment by setton** [comment]
I can reproduce, thank you for the reproducer!

**Comment by reznikmm** [comment]
@c-grant This has been fixed now. I've just published vscode extension 21.0.5 with the fix.

Feel free to report any other issues.

**Comment by Lucretia** [comment]
Just upgraded.

**Comment by oscarrutt** [comment]
I'm using the Ada Language Server in Visual Studio Code on macOS.

This problem is hitting hard in 22.0.999 (and earlier). Sometimes I can work for many minutes without a crash but sometimes, as now, I get a crash almost as soon as I move the cursor into the editing area of a source file. Repeated crashes one after another until I think five crashes and then the server does not start again. It's hard to figure out what to do to make it happen. I can enter the window from the left side and hover over a comment without a crash. But I think it crashes when hovering over _some_ variable names, possibly on the RHS of an assignment statement (wild guess here). This situation is making it impossible to use this language server.

Here is the entire output after the first crash after restarting VSC.

dyld: lazy symbol binding failed: Symbol not found: ____chkstk_darwin
  Referenced from: /Users/me/.vscode/extensions/adacore.ada-22.0.999/darwin/libgmp.10.dylib
  Expected in: /usr/lib/libSystem.B.dylib

dyld: Symbol not found: ____chkstk_darwin
  Referenced from: /Users/me/.vscode/extensions/adacore.ada-22.0.999/darwin/libgmp.10.dylib
  Expected in: /usr/lib/libSystem.B.dylib

[Info  - 3:41:41 AM] Connection to server got closed. Server will restart.
[Error - 3:41:41 AM] Request textDocument/hover failed.
Error: Connection got disposed.
	at Object.dispose (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-jsonrpc/lib/common/connection.js:1009:27)
	at Object.dispose (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-languageclient/lib/common/client.js:71:35)
	at LanguageClient.handleConnectionClosed (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-languageclient/lib/common/client.js:2312:42)
	at LanguageClient.handleConnectionClosed (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-languageclient/lib/node/main.js:146:15)
	at closeHandler (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-languageclient/lib/common/client.js:2299:18)
	at CallbackList.invoke (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-jsonrpc/lib/common/events.js:54:39)
	at Emitter.fire (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-jsonrpc/lib/common/events.js:116:36)
	at closeHandler (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-jsonrpc/lib/common/connection.js:255:26)
	at CallbackList.invoke (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-jsonrpc/lib/common/events.js:54:39)
	at Emitter.fire (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-jsonrpc/lib/common/events.js:116:36)
	at StreamMessageReader.fireClose (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-jsonrpc/lib/common/messageReader.js:39:27)
	at Socket.<anonymous> (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-jsonrpc/lib/common/messageReader.js:124:42)
	at Socket.emit (events.js:327:22)
	at Pipe.<anonymous> (net.js:674:12)
dyld: lazy symbol binding failed: Symbol not found: ____chkstk_darwin
  Referenced from: /Users/me/.vscode/extensions/adacore.ada-22.0.999/darwin/libgmp.10.dylib
  Expected in: /usr/lib/libSystem.B.dylib

dyld: Symbol not found: ____chkstk_darwin
  Referenced from: /Users/me/.vscode/extensions/adacore.ada-22.0.999/darwin/libgmp.10.dylib
  Expected in: /usr/lib/libSystem.B.dylib

[Info  - 3:41:47 AM] Connection to server got closed. Server will restart.
[Error - 3:41:47 AM] Request textDocument/definition failed.
Error: Connection got disposed.
	at Object.dispose (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-jsonrpc/lib/common/connection.js:1009:27)
	at Object.dispose (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-languageclient/lib/common/client.js:71:35)
	at LanguageClient.handleConnectionClosed (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-languageclient/lib/common/client.js:2312:42)
	at LanguageClient.handleConnectionClosed (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-languageclient/lib/node/main.js:146:15)
	at closeHandler (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-languageclient/lib/common/client.js:2299:18)
	at CallbackList.invoke (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-jsonrpc/lib/common/events.js:54:39)
	at Emitter.fire (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-jsonrpc/lib/common/events.js:116:36)
	at closeHandler (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-jsonrpc/lib/common/connection.js:255:26)
	at CallbackList.invoke (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-jsonrpc/lib/common/events.js:54:39)
	at Emitter.fire (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-jsonrpc/lib/common/events.js:116:36)
	at StreamMessageWriter.fireClose (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-jsonrpc/lib/common/messageWriter.js:41:27)
	at Socket.<anonymous> (/Users/me/.vscode/extensions/adacore.ada-22.0.999/node_modules/vscode-jsonrpc/lib/common/messageWriter.js:74:42)
	at Socket.emit (events.js:315:20)
	at Pipe.<anonymous> (net.js:674:12)

**Comment by oscarrutt** [comment]
Here is the macOS crash log from a subsequent restart-and-crash sequence. i've pasted the VSC output above the macOS info but I think it is the same or similar as pasted above in my first post
[VSD output plus macOS crash log.txt](https://github.com/AdaCore/ada_language_server/files/6083006/VSD.output.plus.macOS.crash.log.txt)
.

**Comment by oscarrutt** [comment]
Sorry for multiple posts. Here is the file that I am trying to edit. The crash is repeatable to about 90+% of the time when hovering over d_Theta in line 67 but crashes appear at other locations. Many times just moving the cursor around over the window causes a crash and I don't even know where it was when the crash was triggered. (File renamed to humor github.)
[Antenna_Weighted_Projections_Dissertation.txt](https://github.com/AdaCore/ada_language_server/files/6083097/Antenna_Weighted_Projections_Dissertation.txt)

**Comment by oscarrutt** [comment]
Here's an MWE. Hovering over parts of lines 7 and 8 induces multiple crashes every time. Same when attempting to edit those lines. Removing (1) from range stops the crash. There are also macOS crash logs shown for the extension and sometimes VSC itself crashes. Same behavior for Real_Matrix. In the original file that I posted earlier many many crashes would occur as the cursor is moved around the file's window. It was not always possible to discern what caused the crash.

I endured as many as 200 crashes while I edited down to this extremely short MWE so I hope this problem can be resolved soon.

Thanks for your effort. Here's the program:

with Ada.Numerics.Long_Real_Arrays;
use  Ada.Numerics.Long_Real_Arrays;
procedure dumpme is   
    f : Real_Vector(1 .. 2);  
    b : Integer;
begin
    for i in f'range(1) loop -- Removing (1) stops the crash.
        b := i;
        -- f(i) := Long_Float(i); -- Removes the warning but not the crash.
    end loop;
end dumpme;
